source/material_model/rheology/friction_options.cc

```diff
@@ -0,0 +1,849 @@
+        // Equation (7) from Sobolev and Muldashev (2017):
+        // theta_{n+1} = L/V_{n+1} + (theta_n - L/V_{n+1})*exp(-(V_{n+1}dt)/L)
+        // This is obtained from Equation (5): dtheta/dt = 1 - (theta V)/L
+        // by integration using the assumption that velocities are constant at any time step.
+        double current_theta = critical_slip_distance / ( cellsize * current_edot_ii ) +
+                               (theta_old - critical_slip_distance / ( cellsize * current_edot_ii))
+                               * exp( - ((current_edot_ii * cellsize) * this->get_timestep()) / critical_slip_distance);
+
+        // TODO: make dt the min theta?
+        if (current_theta < 1e-50)
+          current_theta = 1e-50;
+        return current_theta;
+      }
+
+
+
+      template <int dim>
+      void
+      FrictionOptions<dim>::
+      compute_theta_reaction_terms(const int q,
+                                   const std::vector<double> &volume_fractions,
+                                   const MaterialModel::MaterialModelInputs<dim> &in,
+                                   const double min_strain_rate,
+                                   const double ref_strain_rate,
+                                   const bool use_elasticity,
+                                   const bool use_reference_strainrate,
+                                   const double &average_elastic_shear_moduli,
+                                   const double dte,
+                                   MaterialModel::MaterialModelOutputs<dim> &out) const
+      {
+        // Cellsize is needed for theta and the friction angle
+        double cellsize = 1.;
+        if (in.current_cell.state() == IteratorState::valid)
+          {
+            cellsize = in.current_cell->extent_in_direction(0);
+          }
+
+        if (this->simulator_is_past_initialization() && this->get_timestep_number() > 0
+            && in.requests_property(MaterialProperties::reaction_terms)
+            && in.current_cell.state() == IteratorState::valid)
+          {
+            // q is from a for-loop through n_evaluation_points
+            const double current_edot_ii =
+              MaterialUtilities::compute_current_edot_ii (in.composition[q], ref_strain_rate,
+                                                          min_strain_rate, in.strain_rate[q],
+                                                          average_elastic_shear_moduli, use_elasticity,
+                                                          use_reference_strainrate, dte);
+
+            const double theta_old = in.composition[q][theta_composition_index];
+            double current_theta = 0;
+
+            for (unsigned int j=0; j < volume_fractions.size(); ++j)
+              {
+                const double critical_slip_distance = get_critical_slip_distance(in.position[q], j);
+                current_theta += volume_fractions[j] * compute_theta(theta_old, current_edot_ii, cellsize, critical_slip_distance);
```

Comment:
I did that now, because I agree that it makes more sense / feels better like this. However, I think initially I put theta into the lop as well, to ave the theta in reaction terms as similar as possible to the theta used to compute the friction angle. In the latter case everything is within such a loop over the volume fractions within calculate_isostrain_viscosities.

Reply:
Ah that's also true. In that case keeping it the same as calculate_isostrain_visocities is probably better.

In case you don't change it back, there is no need to declare current_friction before the loop and then add to it; you can do `const double current_friction = compute_theta(...);` after the loop.